streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java

```diff
@@ -176,50 +196,80 @@ public boolean isOpen() {
     }
 
     @Override
-    public synchronized long approximateNumEntries() {
+    public long approximateNumEntries() {
         validateStoreOpen();
-        return underlying.approximateNumEntries();
+        lock.readLock().lock();
+        try {
+            return underlying.approximateNumEntries();
+        } finally {
+            lock.readLock().unlock();
+        }
     }
 
     @Override
-    public synchronized void put(final Bytes key, final byte[] value) {
+    public void put(final Bytes key, final byte[] value) {
         Objects.requireNonNull(key, "key cannot be null");
         validateStoreOpen();
-        putInternal(key, value);
+        lock.writeLock().lock();
+        try {
+            putInternal(key, value);
+        } finally {
+            lock.writeLock().unlock();
+        }
     }
 
-    private synchronized void putInternal(final Bytes rawKey, final byte[] value) {
+    private void putInternal(final Bytes rawKey, final byte[] value) {
         Objects.requireNonNull(rawKey, "key cannot be null");
-        cache.put(cacheName, rawKey, new LRUCacheEntry(value, true, context.offset(),
+        lock.writeLock().lock();
```

Comment:
Do we need to lock here? I think the lock has already been taken out from the callers of this method